File: modules/yandex-map/android/src/main/java/expo/modules/yandexmap/model/MarkerModel.kt
package expo.modules.yandexmap.model

import android.graphics.PointF
import com.yandex.mapkit.geometry.Point
import com.yandex.mapkit.map.IconStyle
import com.yandex.mapkit.map.Rect
import com.yandex.mapkit.map.RotationType
import com.yandex.runtime.image.ImageProvider
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

class Coordinate : Record {
    @Field
    val latitude: Double = 0.0

    @Field
    val longitude: Double = 0.0

    fun toPoint(): Point {
        return Point(latitude, longitude)
    }
}

class MarkerIconStyleModel : Record {
    @Field
    val anchor: PointXY? = null

    @Field
    val rotationType: Int? = null

    @Field
    val zIndex: Float = 0.0F

    @Field
    val flat: Boolean = false

    @Field
    val visible: Boolean = true

    @Field
    val scale: Float = 1.0F

    @Field
    val tappableArea: RectRecord? = null

    fun toIconStyle(): IconStyle {
        val rotationTypeEntry = rotationType?.let { RotationType.entries[it] }
        val tappableAreaEntry = tappableArea?.toRect()

        return IconStyle(
            PointF(anchor?.x ?: 1.0f, anchor?.y ?: 1.0f),
            rotationTypeEntry,
            zIndex,
            flat,
            visible,
            scale,
            tappableAreaEntry
        )
    }
}

class PointXY : Record {
    @Field
    var x: Float = 1.0f

    @Field
    var y: Float = 1.0f
}

class RectRecord : Record {
    @Field
    val min: PointXY? = null

    @Field
    val max: PointXY? = null

    fun toRect(): Rect {
        val minPoint = PointF(min?.x ?: 10.0f, min?.y ?: 10.0f)
        val maxPoint = PointF(max?.x ?: 10.0f, max?.y ?: 10.0f)

        return Rect(minPoint, maxPoint)
    }
}

data class PlacemarkConfig(
    var coordinate: Point = Point(0.0, 0.0),
    var iconStyle: IconStyle = IconStyle(),
    var imageProvider: ImageProvider? = null,
)

File: modules/yandex-map/ios/Model/MarkerModel.swift
import ExpoModulesCore
import YandexMapsMobile

struct Coordinate: Record {
    @Field
    var latitude: Double = 0.0

    @Field
    var longitude: Double = 0.0

    func toPoint() -> YMKPoint {
        return YMKPoint(latitude: latitude, longitude: longitude)
    }
}

struct MarkerIconStyleModel: Record {
    @Field
    var anchor: PointXY = PointXY(x: 0.5, y: 0.5)

    @Field
    var rotationType: Float = .zero

    @Field
    var zIndex: Int = .zero

    @Field
    var flat: Bool = false

    @Field
    var visible: Bool = true

    @Field
    var scale: Float = 1.0

    @Field
    var tappableArea: RectRecord? = nil
    
    func toIconStyle() -> YMKIconStyle {
        let iconStyle = YMKIconStyle()
        iconStyle.anchor = NSValue(cgPoint: CGPoint(x: anchor.x, y: anchor.y))
        iconStyle.rotationType = NSNumber(value: rotationType)
        iconStyle.zIndex = NSNumber(value: zIndex)
        iconStyle.flat = NSNumber(value: flat)
        iconStyle.visible = NSNumber(value: visible)
        iconStyle.scale = NSNumber(value: scale)
        iconStyle.tappableArea = tappableArea?.toRect()
        
        return iconStyle
    }
    
}

struct PointXY: Record {
    @Field
    var x: Double = 0.0

    @Field
    var y: Double = 0.0
}

struct RectRecord: Record {
    @Field
    var min: PointXY? = nil

    @Field
    var max: PointXY? = nil

    func toRect() -> YMKRect {
        let minPoint = CGPoint(x: min?.x ?? 10.0, y: min?.y ?? 10.0)
        let maxPoint = CGPoint(x: max?.x ?? 10.0, y: max?.y ?? 10.0)

        return YMKRect(min: minPoint, max: maxPoint)
    }
}

struct PlacemarkConfig {
    var coordinate: YMKPoint = YMKPoint()
    var iconStyle: YMKIconStyle = YMKIconStyle()
    var iconView: UIView = UIView()
}
